File: lua/scratchcards.lua
local aoc = require"adventofcode"
    
function list_of_nums(line)
    return aoc.map(aoc.findall(line, "%d+"), tonumber)
end

function card(line)
    local card_num, left, right = line:match("Card *(%d+):([^|]+)|(.*)")
    return tonumber(card_num), list_of_nums(left), list_of_nums(right)
end

local CARDS = {}

function scratchcards1(line)
    local card_num, winning, drawn = card(line)
    local matches = aoc.intersect(winning, drawn)
    return #matches > 0 and math.tointeger(2^(#matches-1)) or 0
end

function scratchcards2(line)
    local card_num, winning, drawn = card(line)
    CARDS[card_num] = (CARDS[card_num] or 0) + 1
    local matches = aoc.intersect(winning, drawn)
    if #matches > 0 then
        for win_card = card_num+1, card_num+#matches do
            CARDS[win_card] = (CARDS[win_card] or 0) + CARDS[card_num]
        end
    end
    return CARDS[card_num]
end

-- main
aoc.run_reduce_sum({scratchcards1, scratchcards2})

File: python/scratchcards.py
# Advent of Code 2023
# Day 4 task: Scratchcards

from adventofcode import AdventOfCode
import re
from collections import Counter

CARD = re.compile(r"Card *(\d+):([^|]*)\|(.*)")

class scratchcards1(AdventOfCode):

    def line(self, line):
        card,winning,drawn = self.card(line)
        matches = set(winning).intersection(drawn)
        return 2**(len(matches)-1) if matches else 0

    def card(self, line):
        match = CARD.match(line)
        if not match:
            raise ValueError("invalid input")
        list_of_nums = lambda S: map(int,re.findall("\\d+", S))
        return int(match[1]), list_of_nums(match[2]), list_of_nums(match[3])

class scratchcards2(scratchcards1):

    def __init__(self, file):
        self.cardstack = Counter()
        super().__init__(file)

    def line(self, line):
        card,winning,drawn = self.card(line)
        self.cardstack[card] += 1
        matches = set(winning).intersection(drawn)
        if matches:
            # Don't do this, the update method is very slow
            #self.cardstack.update(list(range(card+1, card+len(matches)+1)) * self.cardstack[card])
            for win in range(card+1, card+len(matches)+1):
                self.cardstack[win] += self.cardstack[card]
        return self.cardstack[card]

def main():
    print(AdventOfCode.reduce_and_sum([scratchcards1,scratchcards2]))

if __name__=='__main__':
    main()
